### PR TITLE
Integrate Quac with Qrochet

### LINKED ISSUE #7: Integrate `Quac` with `Qrochet`

### PR DESCRIPTION
**Pull request by Todorbsc**
Closes #7

### CODE REVIEW

**Comment by Todorbsc** [comment]
Two observations:
- Do we need to check if the gates' matrices (tensors) have the right values?
- Is there any case of a quantum circuit where `Biyections` are needed to correctly build the sites?

File: test/runtests.jl [hunk 1]
```diff
@@ -6,6 +6,7 @@ using Qrochet
     include("Quantum_test.jl")
     include("Ansatz/Product_test.jl")
     include("Ansatz/Chain_test.jl")
+    include("integration/QrochetQuacExt_test.jl")
 end
 
 @testset "Integration tests" verbose = true begin end
```
Comment (jofrevalles) [inline test/runtests.jl:12]:
```suggestion
end

@testset "Integration tests" verbose = true begin
    include("integration/QrochetQuacExt_test.jl")
end
```
Nice job! Now, this should be an integration test.

Comment (Todorbsc) [reply test/runtests.jl]: Fixed, thanks!

File: Project.toml [hunk 2]
```diff
@@ -8,7 +8,14 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Tenet = "85d41934-b9cd-44e1-8730-56d86f15f3ec"
 ValSplit = "0625e100-946b-11ec-09cd-6328dd093154"
 
+[weakdeps]
+Quac = "b9105292-1415-45cf-bff1-d6ccf71e6143"
```
[the remark this answers is not in the document]
Comment (Todorbsc) [reply Project.toml:12]: Completely agree. Done!

File: test/integration/QrochetQuacExt_test.jl [hunk 3]
```diff
@@ -0,0 +1,18 @@
+@testset "QrochetQuacExt" begin
+    using Quac
+
+    @testset "QFT_3qubits" begin
```
Comment (mofeing) [inline test/integration/QrochetQuacExt_test.jl:4]: Test set names are for humans to read, so I think this other names are easier.
```suggestion
@testset "Quac" begin
    using Quac

    @testset "QFT" begin
```

File: test/integration/QrochetQuacExt_test.jl [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+        # test sites (inputs and outputs) of the quantum circuit
+        for site in values(qft3qrochet.sites)
+            @test length(qft3qrochet.tn.indexmap[site]) == 1
+        end
+
+        # test inner tensors
+        for notsite in filter(idx -> idx ∉ values(qrqft3.sites), keys(qrqft3.tn.indexmap))
+            @test length(qft3qrochet.tn.indexmap[notsite]) > 1
+        end
```
Comment (mofeing) [inline test/integration/QrochetQuacExt_test.jl:16]: You should try to avoid accessing internal fields and rather use public interfaces like `sites(...)` instead of `.sites`, `TensorNetwork(qtn)` instead of `qtn.tn`...

Also, what are you trying to test in the for loops?

```suggestion
        circuit = Quac.Algorithms.QFT(3)
        qtn = Quantum(circuit)
```

**Review by mofeing — CHANGES_REQUESTED** [review]

File: test/integration/QrochetQuacExt_test.jl [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+        # correct number of inputs and outputs
+        @test ninputs(qftqtn) == n
+        @test noutputs(qftqtn) == n
+        @test socket(qftqtn) == Operator()
+        # all open indices are sites
+        siteinds = getindex.((qftqtn,), sites(qftqtn))
+        @test issetequal(inds(TensorNetwork(qftqtn), :open), siteinds)
+        # all inner indices are not sites
+        notsiteinds = filter(idx -> idx ∉ siteinds, keys(TensorNetwork(qftqtn).indexmap))
```
Comment (mofeing) [inline test/integration/QrochetQuacExt_test.jl:17]: Don't use `.indexmap`. Use `inds(TensorNetwork(qftqtn))`. In this case it's even easier with `setdiff`.
```suggestion
        notsiteinds = setdiff(inds(TensorNetwork(qftqtn)), siteinds)
```

**Review by mofeing — APPROVED** [review]
Yes! This is the kind of tests I was referring to.

I think the "all inner indices are not sites" condition can be too strict, but is ok for now. What do you think @jofrevalles?

LGTM, but just suggested changes.

File: test/integration/QrochetQuacExt_test.jl (hunk 5 above)
Comment (mofeing) [reply test/integration/QrochetQuacExt_test.jl:17]: Actually, I think there is `inds(TensorNetwork(qftqtn), :inner)`. Can you try it?

**Comment by jofrevalles** [comment]
> I think the "all inner indices are not sites" condition can be too strict, but is ok for now. What do you think @jofrevalles?

Yes, I agree that it is too strict, but for now I think this test is OK. I would maybe add a test to see if the contraction of the `TensorNetwork` works and that's it.

**Comment by mofeing** [comment]
> I would maybe add a test to see if the contraction of the `TensorNetwork` works and that's it.

It's better to not mix tests: That's a test for `contract(::TensorNetwork)`, not for `Quantum`.

File: test/integration/QrochetQuacExt_test.jl [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        @test noutputs(qftqtn) == n
+        @test socket(qftqtn) == Operator()
+
+        # all open indices are sites
+        siteinds = getindex.((qftqtn,), sites(qftqtn))
+        @test issetequal(inds(TensorNetwork(qftqtn), :open), siteinds)
+
+        # all inner indices are not sites
+        notsiteinds = setdiff(inds(TensorNetwork(qftqtn)), siteinds)
+        @test issetequal(inds(TensorNetwork(qftqtn), :inner), notsiteinds)
```
Comment (mofeing) [inline test/integration/QrochetQuacExt_test.jl:20]: Let's skip this for now...
```suggestion
        @test_skip issetequal(inds(TensorNetwork(qftqtn), :inner), notsiteinds)
```

File: test/integration/QrochetQuacExt_test.jl [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        # correct number of inputs and outputs
+        @test ninputs(qftqtn) == n
+        @test noutputs(qftqtn) == n
+        @test socket(qftqtn) == Operator()
+
+        # all open indices are sites
+        siteinds = getindex.((qftqtn,), sites(qftqtn))
+        @test issetequal(inds(TensorNetwork(qftqtn), :open), siteinds)
+
+        # all inner indices are not sites
```
Comment (mofeing) [inline test/integration/QrochetQuacExt_test.jl:18]:
```suggestion
        # all inner indices are not sites
        # TODO too strict condition. remove?
```

**Comment by mofeing** [comment]
> * Do we need to check if the gates' matrices (tensors) have the right values?
No, that is the responsability of `Quac`.

> * Is there any case of a quantum circuit where `Biyections` are needed to correctly build the sites?
I don't understand the question. Can you rephrase it?

> * Should we add a test with "Swap" quantum gates?
I don't think we need a special test for that. Do you have an edge case in mind?

**Comment by Todorbsc** [comment]
> > * Do we need to check if the gates' matrices (tensors) have the right values?
> 
> No, that is the responsability of `Quac`.
Perfect!
> > * Is there any case of a quantum circuit where `Biyections` are needed to correctly build the sites?
> 
> I don't understand the question. Can you rephrase it?
In Tenet, `Biyections` were used to find the sites (which was called `interlayer`) so I thought maybe that was necessary (https://github.com/bsc-quantic/Tenet.jl/blob/v0.3.0/ext/TenetQuacExt.jl#L35)
> > * Should we add a test with "Swap" quantum gates?
> 
> I don't think we need a special test for that. Do you have an edge case in mind?
Well, I suggested a test for a quantum circuit with swap gates just to have a complete code coverage of the extension.

**Comment by mofeing** [comment]
> In Tenet, `Biyections` were used to find the sites (which was called `interlayer`) so I thought maybe that was necessary (https://github.com/bsc-quantic/Tenet.jl/blob/v0.3.0/ext/TenetQuacExt.jl#L35)

Nope, that was another implementation.

> Well, I suggested a test for a quantum circuit with swap gates just to have a complete code coverage of the extension.

Well, it's just another gate. Nothing special should happen. If you want to translate the Swap gate to a real swap of the tensor indices, then that is a `Transformation` and so, it's not the responsability of `Quantum`.
